### PR TITLE
Add task ID pattern to notification skip criteria

### LINKED ISSUE #953: Add a task ID specification to the silenced slack notifications information
## Problem

We have some DAGs that may fail at different tasks for different reasons, but with the same predicate. For instance, the Finnish DAG is failing with `AirflowTaskTimeout`s for both the **pull data** task ([reason](https://github.com/WordPress/openverse-catalog/issues/913)) and the **clean data** task ([reason](https://github.com/WordPress/openverse-catalog/issues/450)). We need some way to distinguish these alerts so they can be handled & silenced separately.

## Description

I propose adding a `task_id_pattern` field to the [`SilencedSlackNotification`](https://github.com/WordPress/openverse-catalog/blob/06335539a187036435e0cb02fa94ee4324a7c8da/openverse_catalog/dags/common/slack.py#L69) typed dict which accepts a regex pattern to match against a given task ID (this can't be a direct task ID equality match because the task IDs within task groups _within reingestion workflows_ become quite unique). This would allow us to specify multiple GitHub issues for the same exception across different tasks within a DAG.

### PR DESCRIPTION
**Pull request by AetherUnbound**
## Fixes

Fixes #953 by @AetherUnbound

## Description

This PR adds an _optional_ `task_id_pattern` field which can be used in the `SILENCED_SLACK_NOTIFICATIONS` Variable to specify a specific task within the DAG that a message should be silenced from.

This adds `task_id` as an _optional_ parameter to the various `slack.send_*` functions, so that the new behavior is backwards compatible. I tried to add it as a required variable initially but the changeset would have required so many modifications to various DAGs and tasks that it became unnecessarily cumbersome. Instead, this now adds the `task_id` to all exception calls (e.g. the `on_failure_callback` issuances of `send_alert`), which should cover the error cases we usually want to silence at least. We're fortunate that the silencing infrastructure is so versatile that _any_ message which attempts to get emitted to Slack can be silenced, the drawback is that we can't change the fundamentals of the functions without then changing all of those uses.

I also made some testing tweaks to make it clearer which parameter sets were failing in a given test.

Lastly, I updated the `check_silenced_dags` DAG to include the task ID pattern if it exists.

## Testing Instructions

`just test` will show all of the new tests passing successfully.

To see how the new pattern is available, you can try running the `check_silenced_dag` DAG:

1. Add a new `SILENCED_SLACK_NOTIFICATIONS` variable with the following contents:

```json
{
   "check_silenced_dags": [
        {
            "issue": "https://github.com/WordPress/openverse-catalog/issues/936", 
            "predicate": "ValueError",
            "task_id_pattern": "check_silenced_dags_configuration"
        }
    ]
}
```

2. Run the DAG (you may need the GitHub PAT in order to have this run successfully)
3. Check the logs of the task to see a message similar to the following:

```
[2023-01-10, 23:37:53 UTC] {python.py:177} INFO - Done. Returned value was: The following DAGs have Slack messages silenced, but the associated issue is closed. Please remove them from the silenced_slack_notifications Airflow variable or assign a new issue.
  - <https://github.com/WordPress/openverse-catalog/issues/936|check_silenced_dags (check_silenced_dags_configuration): 'ValueError'>
```

4. Make the following edit to the `check_silenced_dags.py` file:

```diff
diff --git a/openverse_catalog/dags/maintenance/check_silenced_dags.py b/openverse_catalog/dags/maintenance/check_silenced_dags.py
index 17b3277a..b0c8384b 100644
--- a/openverse_catalog/dags/maintenance/check_silenced_dags.py
+++ b/openverse_catalog/dags/maintenance/check_silenced_dags.py
@@ -74,6 +74,7 @@ def get_dags_with_closed_issues(
 
 
 def check_configuration(github_pat: str):
+    raise ValueError("whoops!")
     silenced_dags = Variable.get(
         "SILENCED_SLACK_NOTIFICATIONS", default_var={}, deserialize_json=True
     )
```

5. Run the DAG and check the logs to make sure the following line is present:

```
[2023-01-11, 00:28:54 UTC] {slack.py:294} INFO - Skipping Slack notification for check_silenced_dags::check_silenced_dags_configuration.
```

6. Change the exception to a `KeyError`
7. Run the DAG again and check the logs to make sure "skipped slack message" line was issued (meaning it _would_ have sent a notification)

## Checklist

- [x] My pull request has a descriptive title (not a vague title like
      `Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or
      a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible
      errors.
- [x] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: openverse_catalog/dags/common/slack.py [hunk 1]
```diff
@@ -70,14 +72,17 @@ class SilencedSlackNotification(TypedDict):
...
+           is silenced and tracks resolving the problem.
     predicate: Slack notifications whose text or username contain the
                predicate will be silenced. Matching is case-insensitive.
+    task_id_pattern: A regex pattern that matches the task_id of the task
+                     that triggered the notification. (Optional)
     """
 
     issue: str
     predicate: str
+    task_id_pattern: NotRequired[str | None]
```
Comment (obulat) [inline openverse_catalog/dags/common/slack.py:85]: TIL about NotRequired in TypedDict. Nice!

**Review by obulat — APPROVED** [review]
LGTM!

Is there a docs page about how to use Slack in DAGs?

File: openverse_catalog/dags/common/slack.py (hunk 1 above)
Comment (AetherUnbound) [reply openverse_catalog/dags/common/slack.py:85]: It's a newer addition!

**Comment by AetherUnbound** [comment]
> Is there a docs page about how to use Slack in DAGs?

There's not a docs page specifically, but we do have [documentation on how to use the Slack message builder](https://github.com/WordPress/openverse-catalog/blob/06335539a187036435e0cb02fa94ee4324a7c8da/openverse_catalog/dags/common/slack.py#L2). Almost all functions that use the slack module use either `slack.send_message` or `slack.send_alert`. I could add a note about that to the top of the Slack module if that'd be helpful? Or where else would you expect documentation like that to go?

**Comment by krysal** [comment]
Should the `SILENCED_SLACK_NOTIFICATIONS` variable be added to the `.env` file or through the Airflow UI or somewhere else?

**Comment by AetherUnbound** [comment]
> Should the `SILENCED_SLACK_NOTIFICATIONS` variable be added to the `.env` file or through the Airflow UI or somewhere else?

We only use this "as needed" and it's not necessarily something that we'd care to persist constantly on dev, so it's something we'd edit through the UI (specifically the endpoint `/variable/list/`).

**Review by krysal — APPROVED** [review]
Tha changes are thoroughly tested, nice! Thank you for the detailed test example too, it worked like expected!

It logged the slack message in the step 5. So it just not send the notification, right? I also got this:

```
[2023-02-02, 01:45:49 UTC] {slack.py:317} INFO - 
*DAG*: `check_silenced_dags`
*Task*: `check_silenced_dags_configuration`
*Execution Date*: 2023-02-02T01:43:45Z
*Log*: http://localhost:8080/log?execution_date=2023-02-02T01%3A43%3A45.295353%2B00%3A00&task_id=check_silenced_dags_configuration&dag_id=check_silenced_dags&map_index=-1

*Exception*: whoops!
*Exception Type*: `builtins.ValueError`
```

**Comment by AetherUnbound** [comment]
Yes, it should log the contents of the message regardless of whether the alert gets skipped!
https://github.com/WordPress/openverse-catalog/blob/af205795b412da9f3d0d7f6d7e066b7d039b1763/openverse_catalog/dags/common/slack.py#L295

**Pull request merged**